File: sort/bucket_sort/go/bucket_sort.go
package main

import (
	"math"
	"sort"
)

func bucketSort(arr []int) {
	sqrtN := math.Sqrt(float64(len(arr)))
	n := int(sqrtN)

	buckets := make([][]int, n)
	for idx := range buckets {
		buckets[idx] = []int{}
	}

	max := findMax(arr)
	for _, ele := range arr {
		bIdx := hash(ele, max, n)
		buckets[bIdx] = append(buckets[bIdx], ele)
	}

	arrIdx := 0
	for idx := range buckets {
		sort.Ints(buckets[idx])
		for _, val := range buckets[idx] {
			arr[arrIdx] = val
			arrIdx++
		}
	}
}

func findMax(arr []int) int {
	max := arr[0]

	for i := 1; i < len(arr); i++ {
		if max < arr[i] {
			max = arr[i]
		}
	}

	return max
}

func hash(val, max, n int) int {
	return int(float64(val) / float64(max) * float64(n-1))
}

func main() {

}

File: sort/bucket_sort/python/bucket sort.py
def bucketsort( A ):
  # get hash codes
  code = hashing( A )
  buckets = [list() for _ in range( code[1] )]
  # distribute data into buckets: O(n)
  for i in A:
    x = re_hashing( i, code )
    buck = buckets[x]
    buck.append( i )
  for bucket in buckets:
    bucketsort( bucket )
 
  ndx = 0
  # merge the buckets: O(n)
  for b in range( len( buckets ) ):
    for v in buckets[b]:
      A[ndx] = v
      ndx += 1
 
import math
 
def hashing( A ):
  m = A[0]
  for i in range( 1, len( A ) ):
    if ( m < A[i] ):
      m = A[i]
  result = [m, int( math.sqrt( len( A ) ) )]
  return result
 
 
def re_hashing( i, code ):
  return int( i / code[0] * ( code[1] - 1 ) )
